lang/LangPrimSource/SC_ComPort.h

```diff
@@ -20,103 +20,122 @@
-        socket(ioService),
-        mParent(parent) {}
+class UDP : private boost::noncopyable {
+public:
+    UDP(int inPortNum, HandlerType, int portsToCheck = 10);
+    ~UDP() = default;
 
-    void start();
+    auto RealPortNum() const { return mPortNum; }
+    auto& getSocket() { return mUdpSocket; }
 
 private:
-    int32 OSCMsgLength;
-    char* data;
-    class SC_TcpInPort* mParent;
+    void initHandler(HandlerType type);
 
-    void handleLengthReceived(const boost::system::error_code& error, size_t bytes_transferred);
+    void handleReceivedUDP(const boost::system::error_code& error, std::size_t bytes_transferred);
+    void startReceiveUDP();
 
-    void handleMsgReceived(const boost::system::error_code& error, size_t bytes_transferred);
+    int mPortNum;
+    HandleDataFunc mHandleFunc;
+    boost::array<char, kTextBufSize> mRecvBuffer;
+    boost::asio::ip::udp::endpoint mRemoteEndpoint;
+    boost::asio::ip::udp::socket mUdpSocket;
 };
 
-class SC_TcpInPort {
-    boost::asio::ip::tcp::acceptor acceptor;
-
+class UDPCustom : public UDP {
 public:
-    const int mPortNum;
+    UDPCustom(int inPortNum, HandlerType);
+    ~UDPCustom() = default;
+};
 
-    SC_TcpInPort(int inPortNum, int inMaxConnections, int inBacklog);
+class TCP : private boost::noncopyable {
+public:
+    TCP(int inPortNum, int inMaxConnections, int inBacklog, HandlerType);
 
+private:
     void startAccept();
-    void handleAccept(SC_TcpConnection::pointer new_connection, const boost::system::error_code& error);
+    void handleAccept(Detail::TCPConnection::pointer new_connection, const boost::system::error_code& error);
+
+    HandleDataFunc mHandleFunc;
+    const HandlerType mHandlerType;
+    const int mPortNum;
+    boost::asio::ip::tcp::acceptor mAcceptor;
 };
 
+} // namespace InPort
 
 //////////////////////////////////////////////////////////////////////////////////////////////////////////
 
+namespace OutPort {
 
-class SC_TcpClientPort {
+class TCP : private boost::noncopyable {
 public:
     typedef void (*ClientNotifyFunc)(void* clientData);
 
 public:
-    SC_TcpClientPort(unsigned long inAddress, int inPort, ClientNotifyFunc notifyFunc = 0, void* clientData = 0);
+    TCP(unsigned long inAddress, int inPort, HandlerType, ClientNotifyFunc notifyFunc = 0, void* clientData = 0);
     int Close();
 
-    boost::asio::ip::tcp::socket& Socket() { return socket; }
+    boost::asio::ip::tcp::socket& Socket() { return mSocket; }
 
 private:
-    int32 OSCMsgLength;
-    char* data;
-
     void startReceive();
     void handleLengthReceived(const boost::system::error_code& error, size_t bytes_transferred);
-
     void handleMsgReceived(const boost::system::error_code& error, size_t bytes_transferred);
+    void initHandler(HandlerType);
 
-    boost::asio::ip::tcp::socket socket;
-    boost::asio::ip::tcp::endpoint endpoint;
-
+    HandleDataFunc mHandleFunc;
+    int32 mOSCMsgLength;
+    std::unique_ptr<char[]> mData;
```

Comment:
why does TCP use a `std::unique_ptr<char[]>` but UDP use a `boost::array`?

Reply:
I wanted to keep existing code as close as possible to the original behavior / minimize possibilities for subtle change  since this has very little test coverage. This can definitely be changed to std::array, but probably it should be done in a separate change and tested as such. Maybe I'm just being extra paranoid. :)